hal/halx86/apic/apictimer.c

```diff
@@ -63,22 +69,75 @@ VOID
 NTAPI
 HalStartProfileInterrupt(IN KPROFILE_SOURCE ProfileSource)
 {
-    UNIMPLEMENTED;
-    return;
+	LVT_REGISTER LvtEntry;
+
+	/* Only handle ProfileTime */
+	if (ProfileSource == ProfileTime)
+	{
+		/* OK, we are profiling now */
+		HalIsProfiling = TRUE;
+
+		/* Set interrupt interval */
+		ApicWrite(APIC_TICR, PCR->HalReserved[0]);
+
+		/* Unmask it */
+		LvtEntry.Long = 0;
+		LvtEntry.TimerMode = 1;
+		LvtEntry.Vector = APIC_PROFILE_VECTOR;
+		LvtEntry.Mask = 0;
+		ApicWrite(APIC_TMRLVTR, LvtEntry.Long);
+	}
 }
 
 VOID
 NTAPI
 HalStopProfileInterrupt(IN KPROFILE_SOURCE ProfileSource)
 {
-    UNIMPLEMENTED;
-    return;
+	LVT_REGISTER LvtEntry;
+
+	/* Only handle ProfileTime */
+	if (ProfileSource == ProfileTime)
+	{
+		/* We are not profiling */
+		HalIsProfiling = FALSE;
+
+		/* Mask interrupt */
+		LvtEntry.Long = 0;
+		LvtEntry.TimerMode = 1;
+		LvtEntry.Vector = APIC_PROFILE_VECTOR;
+		LvtEntry.Mask = 1;
+		ApicWrite(APIC_TMRLVTR, LvtEntry.Long);
+	}
 }
 
 ULONG_PTR
 NTAPI
 HalSetProfileInterval(IN ULONG_PTR Interval)
 {
-    UNIMPLEMENTED;
-    return Interval;
+	ULONGLONG TimerInterval;
+	ULONGLONG FixedInterval;
+
+	FixedInterval = (ULONGLONG)Interval;
+
+	/* Check bounds */
+	if (FixedInterval < HalMinProfileInterval)
+		FixedInterval = HalMinProfileInterval;
+	else if (FixedInterval > HalMaxProfileInterval)
+		FixedInterval = HalMaxProfileInterval;
+
+	/* Remember interval */
+	HalCurProfileInterval = FixedInterval;
+
+	/* Set multiplier to 1 */
+	PCR->HalReserved[1] = 1;
+
+	/* Recalculate interval for APIC */
+	TimerInterval = HalpCpuClockFrequency.QuadPart * FixedInterval * PCR->HalReserved[1] / 1000000;
```

Comment:
ApicSetTimerInterval makes a similar calculation but does not use HalReserved:
https://github.com/reactos/reactos/blob/c2c66aff7dacc62d125f2cd61d1167e9a2aa3fd6/hal/halx86/apic/apictimer.c#L29

Should one of them be adjusted to match the other?

Reply:
This formula should be written as:

TimerInterval = FixedInterval * PCR->HalReserved[1] / 1000000;

copy-paste problem, sorry